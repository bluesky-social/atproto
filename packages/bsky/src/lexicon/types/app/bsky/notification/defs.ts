/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { type ValidationResult, BlobRef } from '@atproto/lexicon'
import { CID } from 'multiformats/cid'
import { validate as _validate } from '../../../../lexicons'
import {
  type $Typed,
  is$typed as _is$typed,
  type OmitKey,
} from '../../../../util'

const is$typed = _is$typed,
  validate = _validate
const id = 'app.bsky.notification.defs'

export interface RecordDeleted {
  $type?: 'app.bsky.notification.defs#recordDeleted'
}

const hashRecordDeleted = 'recordDeleted'

export function isRecordDeleted<V>(v: V) {
  return is$typed(v, id, hashRecordDeleted)
}

export function validateRecordDeleted<V>(v: V) {
  return validate<RecordDeleted & V>(v, id, hashRecordDeleted)
}
