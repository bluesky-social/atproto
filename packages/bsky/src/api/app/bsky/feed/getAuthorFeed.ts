import { Server } from '../../../../lexicon'
import { FeedKeyset, composeFeed } from '../util/feed'
import { paginate } from '../../../../db/pagination'
import AppContext from '../../../../context'
import { authOptionalVerifier } from '../util'

export default function (server: Server, ctx: AppContext) {
  server.app.bsky.feed.getAuthorFeed({
    auth: authOptionalVerifier,
    handler: async ({ params, auth }) => {
      const { actor, limit, cursor } = params
      const requester = auth.credentials.did
      const db = ctx.db.db
      const { ref } = db.dynamic

      const feedService = ctx.services.feed(ctx.db)

      let did = ''
      if (actor.startsWith('did:')) {
        did = actor
      } else {
        const actorRes = await db
          .selectFrom('actor')
          .select('did')
          .where('handle', '=', actor)
          .executeTakeFirst()
        if (actorRes) {
          did = actorRes?.did
        }
      }

      // @NOTE mutes applied on pds
      let feedItemsQb = feedService
        .selectFeedItemQb()
        .where('originatorDid', '=', did)

      const keyset = new FeedKeyset(
        ref('feed_item.sortAt'),
        ref('feed_item.cid'),
      )

      feedItemsQb = paginate(feedItemsQb, {
        limit,
        cursor,
        keyset,
      })

      const feedItems = await feedItemsQb.execute()
      const feed = await composeFeed(feedService, feedItems, requester)

      return {
        encoding: 'application/json',
        body: {
          feed,
          cursor: keyset.packFromResult(feedItems),
        },
      }
    },
  })
}
