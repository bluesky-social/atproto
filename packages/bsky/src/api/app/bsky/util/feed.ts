import { TimeCidKeyset } from '../../../../db/pagination'
import { FeedRow } from '../../../../services/feed/types'

export enum FeedAlgorithm {
  ReverseChronological = 'reverse-chronological',
}

export class FeedKeyset extends TimeCidKeyset<FeedRow> {
  labelResult(result: FeedRow) {
    return { primary: result.sortAt, secondary: result.cid }
  }
}

// For users with sparse feeds, avoid scanning more than one week for a single page
export const getFeedDateThreshold = (from: string | undefined, days = 1) => {
  const timelineDateThreshold = from ? new Date(from) : new Date()
  timelineDateThreshold.setDate(timelineDateThreshold.getDate() - days)
  return timelineDateThreshold.toISOString()
}
