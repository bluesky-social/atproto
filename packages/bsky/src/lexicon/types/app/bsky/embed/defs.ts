/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { ValidationResult, BlobRef } from '@atproto/lexicon'
import { CID } from 'multiformats/cid'
import { lexicons } from '../../../../lexicons'
import { $Type, is$typed } from '../../../../util'

const id = 'app.bsky.embed.defs'

/** width:height represents an aspect ratio. It may be approximate, and may not correspond to absolute dimensions in any given unit. */
export interface AspectRatio {
  width: number
  height: number
  [k: string]: unknown
}

export function isAspectRatio(
  v: unknown,
): v is AspectRatio & { $type: $Type<'app.bsky.embed.defs', 'aspectRatio'> } {
  return is$typed(v, id, 'aspectRatio')
}

export function validateAspectRatio(v: unknown) {
  return lexicons.validate(
    `${id}#aspectRatio`,
    v,
  ) as ValidationResult<AspectRatio>
}
