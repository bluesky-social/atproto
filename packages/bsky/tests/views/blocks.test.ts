import assert from 'assert'
import { TestNetwork, RecordRef, SeedClient } from '@atproto/dev-env'
import AtpAgent, { AtUri } from '@atproto/api'
import { BlockedActorError } from '@atproto/api/src/client/types/app/bsky/feed/getAuthorFeed'
import { BlockedByActorError } from '@atproto/api/src/client/types/app/bsky/feed/getAuthorFeed'
import { isThreadViewPost } from '@atproto/api/src/client/types/app/bsky/feed/defs'
import {
  isViewRecord as isEmbedViewRecord,
  isViewBlocked as isEmbedViewBlocked,
} from '@atproto/api/src/client/types/app/bsky/embed/record'
import { forSnapshot } from '../_util'
import basicSeed from '../seeds/basic'

describe('pds views with blocking', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let pdsAgent: AtpAgent
  let sc: SeedClient
  let danBlockCarol: { uri: string }
  let aliceReplyToDan: { ref: RecordRef }
  let carolReplyToDan: { ref: RecordRef }

  let alice: string
  let carol: string
  let dan: string
  let danBlockUri: string

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_views_block',
    })
    agent = network.bsky.getClient()
    pdsAgent = network.pds.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    alice = sc.dids.alice
    carol = sc.dids.carol
    dan = sc.dids.dan
    // add follows to ensure blocks work even w follows
    await sc.follow(carol, dan)
    await sc.follow(dan, carol)
    aliceReplyToDan = await sc.reply(
      alice,
      sc.posts[dan][0].ref,
      sc.posts[dan][0].ref,
      'alice replies to dan',
    )
    carolReplyToDan = await sc.reply(
      carol,
      sc.posts[dan][0].ref,
      sc.posts[dan][0].ref,
      'carol replies to dan',
    )
    // dan blocks carol
    danBlockCarol = await pdsAgent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: carol },
      sc.getHeaders(dan),
    )
    danBlockUri = danBlockCarol.uri
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  it('blocks thread post', async () => {
    const { data: threadAlice } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[carol][0].ref.uriStr },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(threadAlice).toEqual({
      thread: {
        $type: 'app.bsky.feed.defs#blockedPost',
        uri: sc.posts[carol][0].ref.uriStr,
        blocked: true,
        author: {
          did: carol,
          viewer: {
            blockedBy: false,
            blocking: danBlockUri,
          },
        },
      },
    })
    const { data: threadCarol } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(threadCarol).toEqual({
      thread: {
        $type: 'app.bsky.feed.defs#blockedPost',
        uri: sc.posts[dan][0].ref.uriStr,
        blocked: true,
        author: {
          did: dan,
          viewer: {
            blockedBy: true,
          },
        },
      },
    })
  })

  it('blocks thread reply', async () => {
    // Contains reply by carol
    const { data: thread } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[alice][1].ref.uriStr },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(forSnapshot(thread)).toMatchSnapshot()
  })

  it('blocks thread parent', async () => {
    // Parent is a post by dan
    const { data: thread } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: aliceReplyToDan.ref.uriStr },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(forSnapshot(thread)).toMatchSnapshot()
  })

  it('blocks record embeds', async () => {
    // Contains a deep embed of carol's post, blocked by dan
    const { data: thread } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 0, uri: sc.posts[alice][2].ref.uriStr },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(forSnapshot(thread)).toMatchSnapshot()
  })

  it('errors on getting author feed', async () => {
    const attempt1 = agent.api.app.bsky.feed.getAuthorFeed(
      { actor: carol },
      { headers: await network.serviceHeaders(dan) },
    )
    await expect(attempt1).rejects.toThrow(BlockedActorError)

    const attempt2 = agent.api.app.bsky.feed.getAuthorFeed(
      { actor: dan },
      { headers: await network.serviceHeaders(carol) },
    )
    await expect(attempt2).rejects.toThrow(BlockedByActorError)
  })

  it('strips blocked users out of getTimeline', async () => {
    const resCarol = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(
      resCarol.data.feed.some((post) => post.post.author.did === dan),
    ).toBeFalsy()

    const resDan = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(
      resDan.data.feed.some((post) => post.post.author.did === carol),
    ).toBeFalsy()
  })

  it('strips blocked users out of getListFeed', async () => {
    const listRef = await sc.createList(alice, 'test list', 'curate')
    await sc.addToList(alice, alice, listRef)
    await sc.addToList(alice, carol, listRef)
    await sc.addToList(alice, dan, listRef)

    const resCarol = await agent.api.app.bsky.feed.getListFeed(
      { list: listRef.uriStr, limit: 100 },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(
      resCarol.data.feed.some((post) => post.post.author.did === dan),
    ).toBeFalsy()

    const resDan = await agent.api.app.bsky.feed.getListFeed(
      { list: listRef.uriStr, limit: 100 },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(
      resDan.data.feed.some((post) => post.post.author.did === carol),
    ).toBeFalsy()
  })

  it('returns block status on getProfile', async () => {
    const resCarol = await agent.api.app.bsky.actor.getProfile(
      { actor: dan },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(resCarol.data.viewer?.blocking).toBeUndefined()
    expect(resCarol.data.viewer?.blockedBy).toBe(true)

    const resDan = await agent.api.app.bsky.actor.getProfile(
      { actor: carol },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(resDan.data.viewer?.blocking).toBeDefined()
    expect(resDan.data.viewer?.blockedBy).toBe(false)
  })

  it('unsets viewer follow state when blocked', async () => {
    // there are follows between carol and dan
    const { data: profile } = await agent.api.app.bsky.actor.getProfile(
      { actor: carol },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(profile.viewer?.following).toBeUndefined()
    expect(profile.viewer?.followedBy).toBeUndefined()
    const { data: result } = await agent.api.app.bsky.graph.getBlocks(
      {},
      { headers: await network.serviceHeaders(dan) },
    )
    const blocked = result.blocks.find((block) => block.did === carol)
    expect(blocked).toBeDefined()
    expect(blocked?.viewer?.following).toBeUndefined()
    expect(blocked?.viewer?.followedBy).toBeUndefined()
  })

  it('returns block status on getProfiles', async () => {
    const resCarol = await agent.api.app.bsky.actor.getProfiles(
      { actors: [alice, dan] },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(resCarol.data.profiles[0].viewer?.blocking).toBeUndefined()
    expect(resCarol.data.profiles[0].viewer?.blockingByList).toBeUndefined()
    expect(resCarol.data.profiles[0].viewer?.blockedBy).toBe(false)
    expect(resCarol.data.profiles[1].viewer?.blocking).toBeUndefined()
    expect(resCarol.data.profiles[1].viewer?.blockingByList).toBeUndefined()
    expect(resCarol.data.profiles[1].viewer?.blockedBy).toBe(true)

    const resDan = await agent.api.app.bsky.actor.getProfiles(
      { actors: [alice, carol] },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(resDan.data.profiles[0].viewer?.blocking).toBeUndefined()
    expect(resDan.data.profiles[0].viewer?.blockingByList).toBeUndefined()
    expect(resDan.data.profiles[0].viewer?.blockedBy).toBe(false)
    expect(resDan.data.profiles[1].viewer?.blocking).toBeDefined()
    expect(resDan.data.profiles[1].viewer?.blockingByList).toBeUndefined()
    expect(resDan.data.profiles[1].viewer?.blockedBy).toBe(false)
  })

  it('does not return block violating follows', async () => {
    const resCarol = await agent.api.app.bsky.graph.getFollows(
      { actor: carol },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(resCarol.data.follows.some((f) => f.did === dan)).toBe(false)

    const resDan = await agent.api.app.bsky.graph.getFollows(
      { actor: dan },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(resDan.data.follows.some((f) => f.did === carol)).toBe(false)
  })

  it('does not return block violating followers', async () => {
    const resCarol = await agent.api.app.bsky.graph.getFollowers(
      { actor: carol },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(resCarol.data.followers.some((f) => f.did === dan)).toBe(false)

    const resDan = await agent.api.app.bsky.graph.getFollowers(
      { actor: dan },
      { headers: await network.serviceHeaders(alice) },
    )
    expect(resDan.data.followers.some((f) => f.did === carol)).toBe(false)
  })

  it('does not return posts from blocked users', async () => {
    const alicePost = sc.posts[alice][0].ref.uriStr
    const carolPost = sc.posts[carol][0].ref.uriStr
    const danPost = sc.posts[dan][0].ref.uriStr

    const resCarol = await agent.api.app.bsky.feed.getPosts(
      { uris: [alicePost, carolPost, danPost] },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(resCarol.data.posts.some((p) => p.uri === alicePost)).toBe(true)
    expect(resCarol.data.posts.some((p) => p.uri === carolPost)).toBe(true)
    expect(resCarol.data.posts.some((p) => p.uri === danPost)).toBe(false)

    const resDan = await agent.api.app.bsky.feed.getPosts(
      { uris: [alicePost, carolPost, danPost] },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(resDan.data.posts.some((p) => p.uri === alicePost)).toBe(true)
    expect(resDan.data.posts.some((p) => p.uri === carolPost)).toBe(false)
    expect(resDan.data.posts.some((p) => p.uri === danPost)).toBe(true)
  })

  it('does not return notifs for blocked accounts', async () => {
    const resCarol = await agent.api.app.bsky.notification.listNotifications(
      {
        limit: 100,
      },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(
      resCarol.data.notifications.some((notif) => notif.author.did === dan),
    ).toBeFalsy()

    const resDan = await agent.api.app.bsky.notification.listNotifications(
      {
        limit: 100,
      },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(
      resDan.data.notifications.some((notif) => notif.author.did === carol),
    ).toBeFalsy()
  })

  it('does not return blocked accounts in actor search', async () => {
    const resCarol = await agent.api.app.bsky.actor.searchActors(
      {
        term: 'dan.test',
      },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(resCarol.data.actors.some((actor) => actor.did === dan)).toBeFalsy()

    const resDan = await agent.api.app.bsky.actor.searchActors(
      {
        term: 'carol.test',
      },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(resDan.data.actors.some((actor) => actor.did === carol)).toBeFalsy()
  })

  it('does not return blocked accounts in actor search typeahead', async () => {
    const resCarol = await agent.api.app.bsky.actor.searchActorsTypeahead(
      {
        term: 'dan.test',
      },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(resCarol.data.actors.some((actor) => actor.did === dan)).toBeFalsy()

    const resDan = await agent.api.app.bsky.actor.searchActorsTypeahead(
      {
        term: 'carol.test',
      },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(resDan.data.actors.some((actor) => actor.did === carol)).toBeFalsy()
  })

  it('does not return blocked accounts in get suggestions', async () => {
    // unfollow so they _would_ show up in suggestions if not for block
    await sc.unfollow(carol, dan)
    await sc.unfollow(dan, carol)
    await network.processAll()

    const resCarol = await agent.api.app.bsky.actor.getSuggestions(
      {
        limit: 100,
      },
      { headers: await network.serviceHeaders(carol) },
    )
    expect(resCarol.data.actors.some((actor) => actor.did === dan)).toBeFalsy()

    const resDan = await agent.api.app.bsky.actor.getSuggestions(
      {
        limit: 100,
      },
      { headers: await network.serviceHeaders(dan) },
    )
    expect(resDan.data.actors.some((actor) => actor.did === carol)).toBeFalsy()
  })

  it('does not serve blocked replies', async () => {
    const getThreadPostUri = (r) => r?.['post']?.['uri']
    // reply then block
    const { data: replyThenBlock } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
        { headers: await network.serviceHeaders(alice) },
      )
    assert(isThreadViewPost(replyThenBlock.thread))
    expect(replyThenBlock.thread.replies?.map(getThreadPostUri)).toEqual([
      aliceReplyToDan.ref.uriStr,
    ])

    // unblock
    await pdsAgent.api.app.bsky.graph.block.delete(
      { repo: dan, rkey: new AtUri(danBlockCarol.uri).rkey },
      sc.getHeaders(dan),
    )
    await network.processAll()
    const { data: unblock } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
      { headers: await network.serviceHeaders(alice) },
    )
    assert(isThreadViewPost(unblock.thread))
    expect(unblock.thread.replies?.map(getThreadPostUri)).toEqual([
      carolReplyToDan.ref.uriStr,
      aliceReplyToDan.ref.uriStr,
    ])

    // block then reply
    danBlockCarol = await pdsAgent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: carol },
      sc.getHeaders(dan),
    )
    const carolReplyToDan2 = await sc.reply(
      carol,
      sc.posts[dan][1].ref,
      sc.posts[dan][1].ref,
      'carol replies to dan again',
    )
    await network.processAll()
    const { data: blockThenReply } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
        { headers: await network.serviceHeaders(alice) },
      )
    assert(isThreadViewPost(blockThenReply.thread))
    expect(replyThenBlock.thread.replies?.map(getThreadPostUri)).toEqual([
      aliceReplyToDan.ref.uriStr,
    ])

    // cleanup
    await pdsAgent.api.app.bsky.feed.post.delete(
      { repo: carol, rkey: carolReplyToDan2.ref.uri.rkey },
      sc.getHeaders(carol),
    )
    await network.processAll()
  })

  it('does not serve blocked embeds to third-party', async () => {
    // embed then block
    const { data: embedThenBlock } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 0, uri: sc.posts[dan][1].ref.uriStr },
        { headers: await network.serviceHeaders(alice) },
      )
    assert(isThreadViewPost(embedThenBlock.thread))
    assert(isEmbedViewBlocked(embedThenBlock.thread.post.embed?.record))

    // unblock
    await pdsAgent.api.app.bsky.graph.block.delete(
      { repo: dan, rkey: new AtUri(danBlockCarol.uri).rkey },
      sc.getHeaders(dan),
    )
    await network.processAll()
    const { data: unblock } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 0, uri: sc.posts[dan][1].ref.uriStr },
      { headers: await network.serviceHeaders(alice) },
    )
    assert(isThreadViewPost(unblock.thread))
    assert(isEmbedViewRecord(unblock.thread.post.embed?.record))

    // block then embed
    danBlockCarol = await pdsAgent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: carol },
      sc.getHeaders(dan),
    )
    const carolEmbedsDan = await sc.post(
      carol,
      'carol embeds dan',
      undefined,
      undefined,
      sc.posts[dan][0].ref,
    )
    await network.processAll()
    const { data: blockThenEmbed } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 0, uri: carolEmbedsDan.ref.uriStr },
        { headers: await network.serviceHeaders(alice) },
      )
    assert(isThreadViewPost(blockThenEmbed.thread))
    assert(isEmbedViewBlocked(blockThenEmbed.thread.post.embed?.record))

    // cleanup
    await pdsAgent.api.app.bsky.feed.post.delete(
      { repo: carol, rkey: carolEmbedsDan.ref.uri.rkey },
      sc.getHeaders(carol),
    )
    await network.processAll()
  })

  it('applies third-party blocking rules in feeds.', async () => {
    // alice follows carol and dan, block exists between carol and dan.
    const replyBlockedUri = carolReplyToDan.ref.uriStr
    const embedBlockedUri = sc.posts[dan][1].ref.uriStr
    const { data: timeline } = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: await network.serviceHeaders(alice) },
    )
    const replyBlockedPost = timeline.feed.find(
      (item) => item.post.uri === replyBlockedUri,
    )
    expect(replyBlockedPost).toBeUndefined()
    const embedBlockedPost = timeline.feed.find(
      (item) => item.post.uri === embedBlockedUri,
    )
    assert(embedBlockedPost)
    assert(isEmbedViewBlocked(embedBlockedPost.post.embed?.record))
  })

  it('returns a list of blocks', async () => {
    await pdsAgent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: alice },
      sc.getHeaders(dan),
    )

    await network.processAll()

    const res = await agent.api.app.bsky.graph.getBlocks(
      {},
      { headers: await network.serviceHeaders(dan) },
    )
    const dids = res.data.blocks.map((block) => block.did).sort()
    expect(dids).toEqual([alice, carol].sort())
  })

  it('paginates getBlocks', async () => {
    const full = await agent.api.app.bsky.graph.getBlocks(
      {},
      { headers: await network.serviceHeaders(dan) },
    )
    const first = await agent.api.app.bsky.graph.getBlocks(
      { limit: 1 },
      { headers: await network.serviceHeaders(dan) },
    )
    const second = await agent.api.app.bsky.graph.getBlocks(
      { cursor: first.data.cursor },
      { headers: await network.serviceHeaders(dan) },
    )
    const combined = [...first.data.blocks, ...second.data.blocks]
    expect(combined).toEqual(full.data.blocks)
  })
})
