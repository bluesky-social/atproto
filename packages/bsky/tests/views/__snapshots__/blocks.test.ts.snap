// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds views with blocking blocks record embeds 1`] = `
Object {
  "thread": Object {
    "$type": "app.bsky.feed.defs#threadViewPost",
    "post": Object {
      "author": Object {
        "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "labels": Array [
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "embed": Object {
        "$type": "app.bsky.embed.record#view",
        "record": Object {
          "$type": "app.bsky.embed.record#viewRecord",
          "author": Object {
            "did": "user(2)",
            "handle": "dan.test",
            "labels": Array [],
            "viewer": Object {
              "blockedBy": false,
              "muted": false,
            },
          },
          "cid": "cids(3)",
          "embeds": Array [
            Object {
              "$type": "app.bsky.embed.record#view",
              "record": Object {
                "$type": "app.bsky.embed.record#viewBlocked",
                "author": Object {
                  "did": "user(3)",
                  "viewer": Object {
                    "blockedBy": false,
                    "blocking": "record(5)",
                  },
                },
                "blocked": true,
                "uri": "record(4)",
              },
            },
          ],
          "indexedAt": "1970-01-01T00:00:00.000Z",
          "labels": Array [],
          "uri": "record(3)",
          "value": Object {
            "$type": "app.bsky.feed.post",
            "createdAt": "1970-01-01T00:00:00.000Z",
            "embed": Object {
              "$type": "app.bsky.embed.record",
              "record": Object {
                "cid": "cids(4)",
                "uri": "record(4)",
              },
            },
            "facets": Array [
              Object {
                "features": Array [
                  Object {
                    "$type": "app.bsky.richtext.facet#mention",
                    "did": "user(0)",
                  },
                ],
                "index": Object {
                  "byteEnd": 18,
                  "byteStart": 0,
                },
              },
            ],
            "text": "@alice.bluesky.xyz is the best",
          },
        },
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [
        Object {
          "cid": "cids(0)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "did:example:labeler",
          "uri": "record(0)",
          "val": "test-label",
        },
      ],
      "likeCount": 2,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "embed": Object {
          "$type": "app.bsky.embed.record",
          "record": Object {
            "cid": "cids(3)",
            "uri": "record(3)",
          },
        },
        "text": "yoohoo label_me",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "record(0)",
      "viewer": Object {},
    },
  },
}
`;

exports[`pds views with blocking blocks thread parent 1`] = `
Object {
  "thread": Object {
    "$type": "app.bsky.feed.defs#threadViewPost",
    "parent": Object {
      "$type": "app.bsky.feed.defs#blockedPost",
      "author": Object {
        "did": "user(2)",
        "viewer": Object {
          "blockedBy": true,
        },
      },
      "blocked": true,
      "uri": "record(4)",
    },
    "post": Object {
      "author": Object {
        "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "labels": Array [
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(3)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(3)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(2)",
          "following": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "reply": Object {
          "parent": Object {
            "cid": "cids(3)",
            "uri": "record(4)",
          },
          "root": Object {
            "cid": "cids(3)",
            "uri": "record(4)",
          },
        },
        "text": "alice replies to dan",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "record(0)",
      "viewer": Object {},
    },
    "replies": Array [],
  },
}
`;

exports[`pds views with blocking blocks thread reply 1`] = `
Object {
  "thread": Object {
    "$type": "app.bsky.feed.defs#threadViewPost",
    "post": Object {
      "author": Object {
        "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "labels": Array [
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 3,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000000Z",
        "text": "again",
      },
      "replyCount": 2,
      "repostCount": 1,
      "uri": "record(0)",
      "viewer": Object {
        "like": "record(4)",
        "repost": "record(3)",
      },
    },
    "replies": Array [
      Object {
        "$type": "app.bsky.feed.defs#blockedPost",
        "author": Object {
          "did": "user(2)",
          "viewer": Object {
            "blockedBy": false,
            "blocking": "record(6)",
          },
        },
        "blocked": true,
        "uri": "record(5)",
      },
      Object {
        "$type": "app.bsky.feed.defs#threadViewPost",
        "post": Object {
          "author": Object {
            "avatar": "https://bsky.public.url/img/avatar/plain/user(4)/cids(1)@jpeg",
            "did": "user(3)",
            "displayName": "bobby",
            "handle": "bob.test",
            "labels": Array [],
            "viewer": Object {
              "blockedBy": false,
              "following": "record(8)",
              "muted": false,
            },
          },
          "cid": "cids(3)",
          "embed": Object {
            "$type": "app.bsky.embed.images#view",
            "images": Array [
              Object {
                "alt": "tests/sample-img/key-landscape-small.jpg",
                "fullsize": "https://bsky.public.url/img/feed_fullsize/plain/user(4)/cids(4)@jpeg",
                "thumb": "https://bsky.public.url/img/feed_thumbnail/plain/user(4)/cids(4)@jpeg",
              },
            ],
          },
          "indexedAt": "1970-01-01T00:00:00.000Z",
          "labels": Array [
            Object {
              "cid": "cids(3)",
              "cts": "1970-01-01T00:00:00.000Z",
              "neg": false,
              "src": "did:example:labeler",
              "uri": "record(7)",
              "val": "test-label",
            },
            Object {
              "cid": "cids(3)",
              "cts": "1970-01-01T00:00:00.000Z",
              "neg": false,
              "src": "did:example:labeler",
              "uri": "record(7)",
              "val": "test-label-2",
            },
          ],
          "likeCount": 0,
          "record": Object {
            "$type": "app.bsky.feed.post",
            "createdAt": "1970-01-01T00:00:00.000Z",
            "embed": Object {
              "$type": "app.bsky.embed.images",
              "images": Array [
                Object {
                  "alt": "tests/sample-img/key-landscape-small.jpg",
                  "image": Object {
                    "$type": "blob",
                    "mimeType": "image/jpeg",
                    "ref": Object {
                      "$link": "cids(4)",
                    },
                    "size": 4114,
                  },
                },
              ],
            },
            "reply": Object {
              "parent": Object {
                "cid": "cids(0)",
                "uri": "record(0)",
              },
              "root": Object {
                "cid": "cids(0)",
                "uri": "record(0)",
              },
            },
            "text": "hear that label_me label_me_2",
          },
          "replyCount": 1,
          "repostCount": 0,
          "uri": "record(7)",
          "viewer": Object {},
        },
      },
    ],
  },
}
`;
